store/aerospike/aerospike.go

```diff
@@ -0,0 +1,322 @@
+	stmt := api.NewStatement(s.namespace, "")
+	stmt.SetAggregateFunction("valkeyrie", "list", []api.Value{api.NewValue(keyPrefix)}, true)
+	rs, err := s.client.Query(nil, stmt)
+	if err != nil {
+		return nil, err
+	}
+
+	kv := []*store.KVPair{}
+
+	keyFound := false
+	for res := range rs.Results() {
+		if res.Err != nil {
+			return nil, res.Err
+		}
+		record := res.Record.Bins["SUCCESS"].(map[interface{}]interface{})
+		if record["key"] != nil {
+			if record["key"] != keyPrefix {
+				kv = append(kv, &store.KVPair{
+					Key:       record["key"].(string),
+					Value:     record["bin"].([]byte),
+					LastIndex: uint64(record["gen"].(int)),
+				})
+			} else {
+				keyFound = true
+			}
+		}
+	}
+
+	if keyFound && len(kv) == 0 {
+		return []*store.KVPair{}, nil
+	}
+	if !keyFound && len(kv) == 0 {
+		return nil, store.ErrKeyNotFound
+	}
+
+	return kv, nil
+}
+
+// DeleteTree deletes a range of keys under a given substring
+// Uses the List - very slowly!
+// It is not recommended to use without special need.
+func (s *Aerospike) DeleteTree(keyFilter string) error {
+	rs, err := s.List(keyFilter, nil)
+	if err != nil {
+		return err
+	}
+	for _, res := range rs {
+		err = s.Delete(res.Key)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// AtomicPut put a value at "key" if the key has not been
+// modified in the meantime, throws an error if this is the case
+func (s *Aerospike) AtomicPut(key string, value []byte, previous *store.KVPair, opts *store.WriteOptions) (bool, *store.KVPair, error) {
+	akey, err := s.newKey(key)
+	if err != nil {
+		return false, nil, err
+	}
+
+	policy := api.NewWritePolicy(0, 0)
+	policy.SendKey = true
+	if opts != nil && opts.TTL > 0 {
+		policy.Expiration = uint32(opts.TTL / time.Second)
+	}
+	if previous != nil {
+		policy.GenerationPolicy = api.EXPECT_GEN_EQUAL
+		policy.Generation = uint32(previous.LastIndex)
+	} else {
+		policy.RecordExistsAction = api.CREATE_ONLY
+	}
+
+	bin := api.NewBin("bin", value)
+	err = s.client.PutBins(policy, akey, bin)
+	if err != nil {
+		if ae, ok := err.(types.AerospikeError); ok && ae.ResultCode() == types.KEY_EXISTS_ERROR {
+			return false, nil, store.ErrKeyExists
+		}
+		return false, nil, store.ErrKeyModified
+	}
+
+	updated := &store.KVPair{
+		Key:       key,
+		Value:     value,
+		LastIndex: uint64(atomic.AddUint32(&policy.Generation, 1)),
+	}
+
+	return true, updated, nil
+}
+
+// AtomicDelete deletes a value at "key" if the key has not
+// been modified in the meantime, throws an error if this is the case
+func (s *Aerospike) AtomicDelete(key string, previous *store.KVPair) (bool, error) {
+	if previous == nil {
+		return false, store.ErrKeyModified
+	}
+
+	// Extra Get operation to check on the key
+	pair, err := s.Get(key, nil)
+	if err != nil {
+		return false, err
+	}
+	if pair.LastIndex != previous.LastIndex {
+		return false, store.ErrKeyModified
+	}
+
+	if err := s.Delete(key); err != nil {
```

Comment:
I'm not so sure about the implementation of the atomic delete as a client could modify the key between the pair index check above and the `Delete`. This would end up with a client issuing a put but having its key being deleted even though it shouldn't. As is it's not really atomic as we do that check on the client but on the DB with linearizable properties. May be wrong on that one but a second lecture will help.

Reply:
there is such a probability, but I do not see other options...

Reply:
did similarly to this implementation https://github.com/abronan/valkeyrie/blob/master/store/boltdb/boltdb.go#L337